ESQL: Fully replace text fields with their exact representation

costin opened the pull request:
Adjust the rule to perform the replacement everywhere to avoid
 attributes with the same id (thus semantically the same) but different
 data types

costin committed:
ESQL: Fully replace text fields with their exact representation

Adjust the rule to perform the replacement everywhere to avoid
 attributes with the same id (thus semantically the same) but different
 data types

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

costin committed:
Add layout check

nik9000 reviewed (approved):
It unblocks me! Any other QL folks want to have a look?

costin committed:
Make sure to use the proper FieldAttribute of the exact subfield instead
 of duplicating the nameId of the inexact field attribute.
The previous approach used the parent name id and changed its data type
 which caused all sorts of problems downstream.  The rule now properly
 checks the FieldAttributes on the source and does the replacement using
 the detected lookup table instead of just reusing the name id.

costin committed:
Massage code

nik9000 commented:
Oh no! It fails the CSV tests. There's another sneaky thing going on. Technically I think this'd unlock me if we dropped that extra test in the Layout.Builder. I don't know that I'd want to do that though. Maybe it's worth it temporarily though.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -856,21 +859,35 @@ protected Expression regexToEquals(RegexMatch<?> regexMatch, Literal literal) {
+                // everything else uses the map to replace the field with the proper attribute
+                else {
+                    p = p.transformExpressionsOnly(FieldAttribute.class, f -> {
+                        if (f.dataType() == DataTypes.TEXT && f.getExactInfo().hasExact()) {
```
Why the `TEXT` check? I mean, I know there are no other field types that use the "exact" functionality, but it's a nice functionality that elegantly abstracts this concept.

astefan replied:
Because there can be other field types with "exact" subfields (keyword parent field with a keyword sub-field) but only `text` has a concrete use of the sub-field.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/Layout.java:
```diff
@@ -118,7 +118,11 @@ public Layout build() {
+                    ChannelAndType next = new ChannelAndType(channel, set.type);
+                    ChannelAndType prev = layout.put(id, next);
+                    if (prev != null) {
+                        throw new IllegalArgumentException("Name [" + id + "] is on two channels [" + prev + "] and [" + next + "]");
```
YES! We needed this.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -856,21 +859,35 @@ protected Expression regexToEquals(RegexMatch<?> regexMatch, Literal literal) {
+                    for (Attribute out : r.output()) {
+                        // map the field to their actual attributes
+                        if (out instanceof FieldAttribute fa) {
+                            esFieldToAttribute.put(fa.field(), fa);
```
question: This code assumes that for each field attribute, there can only be one corresponding `EsField` with that attribute (otherwise, this line would replace the previous field attribute and it would get lost).

Why is that? In the tests that were updated for this PR, I can see that the `job` field has an attribute `job.raw`, which is used in place of `job` after optimization. I wonder if the tests would still work if `job` had additional nested properties `job.prop1`, `job.prop2` etc.

Or, maybe the correspondence `FieldAttribute` -> `EsField` is unique for text fields and we don't care about fields with nested properties etc.? In this case, a small comment here would help understand the assumptions when verifying correctness as a reader.

costin replied:
ES fields are just regular fields that as a convention, use a certain name pattern.
Each one can be addressed individually and can have different data types - from the ES perspective is up to the user to figure the relationship between them and use them as such.
Having an attribute to multiple ES fields would create more problems than it would solve.
In terms of documentation - isn't that obvious in how the relationship is designed? FieldAttribute points to one EsField not a list; what else is needed?

alex-spies replied:
Thanks, I think I just got confused about the exact meaning of the `FieldAttribute` class.

So, what got me confused is that there's fields, there's properties and there's field attributes. It was not obvious to me from the `FieldAttribute` javadoc that a `FieldAttribute` simply represents an `ESField` as an attribute. It could also be that a `FieldAttribute` represents something belonging to an `ESField`, like any of its properties, and then we'd have a many-to-one relationship which would make the code above faulty. (Also, the constructor of `FieldAttribute` does not construct the attribute _from_ an `ESField`, but just takes it and stores it, so the relationship between field attribs and es fields is not obvious from the `FieldAttribute` class itself.)

If you think that could help avoid more misunderstandings in the future, I could put together a quick doc PR that elaborates the `FieldAttribute` javadoc a bit. What do you think?

alex-spies reviewed (approved):
LGTM, only a question regarding uniqueness of the `EsField -> FieldAttribute` map (which shouldn't affect correctness of the overall rule, anyway).

astefan commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizerTests.java:
```diff
@@ -420,7 +419,7 @@ public void testExtractorMultiEvalWithSameName() {
         var extract = as(project.child(), FieldExtractExec.class);
         assertThat(
             names(extract.attributesToExtract()),
-            contains("_meta_field", "emp_no", "first_name", "gender", "job", "job.raw", "languages", "last_name", "salary")
```
I don't think the test is reflecting the reality. The FieldExtractExec here should ask for `job` as well.

astefan replied:
In the end the list of outputted fields/columns includes both `job` and `job.raw` and they can be different. If `raw` is a sub-field of type `keyword` with `ignore_above: 5` it can have no value (nothing is indexed in that field), while its parent does.

nik9000 commented:
Apply that patch gets most everything to pass. the remaining failures are all in the yaml files around names. Which I think just need updating.

costin committed:
Fix formatting

costin committed:
Filter duplicates

costin committed:
Merge branch 'main' into esql/fully-replace-text-w-exact

astefan commented:
> Apply that patch gets most everything to pass. the remaining failures are all in the yaml files around names. Which I think just need updating.

There's an aspect here that I am not sure of: the names of the fields/columns returned do not reflect the query intention anymore. If I use in my query `keep job` I am expecting to see a column called `job`, because there is such a field present in the index. The latest yml failure though indicates that `job` is replaced "everywhere" with `job.raw` (including the final output) and I don't think that's ok. There is one thing to replace a `text` field with its exact subfield everywhere, and another to make the replacement in those places that "matter" (those that do query matching and transformations) and in the end to show the original value.

There is a subtle aspect here that we also encountered in ES SQL: we match using one value, but we show to the user another. On one hand, `job` is what the user wants to see but at that point the value **that actually matched** is different (the value of `job.raw`), or we show a value that's not the one of the `job` field, but its exact representation, and we also use its exact representation in matches/queries.

costin commented:
@astefan great comment.

The text -> keyword fallback is a subtle one and as a language we end up having to clarify the rules around it without breaking the expectation.
Let me step back and restate the problem, how we got here and how we can move forward:.

Problem statement: "text fields are meant for full-text search NOT retrieval".
In other words, if there would be no keyword, the user would have to do source extraction. If there is no source, there's nothing to extract (I'm not considering the analyzed/normalized content).
In _search, this is is much more obvious yet but in ESQL this is abstracted away.
In SQL we are handling this at the extraction level so the attribute/field act as if they're the original yet we know that is not the case.
I'd like that we avoid that in ESQL so we're much transparent about the actual type (and to some degree about the underlying field) and upfront about how extraction is done instead of doing the switcheroo at the extraction level.

I cannot think of any other field that requires this fallback then TEXT hence why I wouldn't look into generalizing hence why the current PR changes the fields *everywhere* in the query instead of just certain places - we want the query to be consistent and having the same `FieldAttribute` with different backing `EsField` depending on the query is NOT that.

costin committed:
Fallback to exact attribute at extraction phase instead of planning

Replacing at the plan level an attribute with its exact subfield ends up
 breaks down when performing extraction since the name is used verbatim.
 To avoid more disruptive changes, use the SQL approach of switching the
 field late in the process, when doing query translation.

costin commented:
@elasticsearchmachine run elasticsearch-ci/part-3

costin committed:
Merge branch 'main' into esql/fully-replace-text-w-exact
